Library/OcCryptoLib/BigNumPrimitives.c

```diff
@@ -0,0 +1,1010 @@
+/**
+  This library performs unsigned arbitrary precision arithmetic operations.
+  All results are returned into caller-provided buffers. The caller is
+  responsible to ensure the buffers can hold a value of the precision it
+  desires. Too large results will be truncated without further notification for
+  public APIs.
+
+  https://github.com/kokke/tiny-bignum-c has served as a template for several
+  algorithmic ideas.
+
+  This code is not to be considered general-purpose but solely to support
+  cryptographic operations such as RSA encryption. As such, there are arbitrary
+  limitations, such as requirement of equal precision, to limit the complexity
+  of the operations to the bare minimum required to support such use caes.
+
+  SECURITY: Currently, no security measures have been taken. This code is
+            vulnerable to both timing and side channel attacks for value
+            leakage. However, its current purpose is the verification of public
+            binaries with public certificates, for which this is perfectly
+            acceptable, especially in regards to performance.
+
+  Copyright (C) 2019, Download-Fritz. All rights reserved.
+
+This program and the accompanying materials
+are licensed and made available under the terms and conditions of the BSD License
+which accompanies this distribution.  The full text of the license may be found at
+http://opensource.org/licenses/bsd-license.php
+
+THE PROGRAM IS DISTRIBUTED UNDER THE BSD LICENSE ON AN "AS IS" BASIS,
+WITHOUT WARRANTIES OR REPRESENTATIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED.
+
+**/
+
+#include <Base.h>
+
+#include <Library/BaseLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/DebugLib.h>
+#include <Library/MemoryAllocationLib.h>
+#include <Library/OcCryptoLib.h>
+#include <Library/OcGuardLib.h>
+
+#include "BigNumLibInternal.h"
+
+#define OC_BN_MAX_VAL  ((OC_BN_WORD)0U - 1U)
+
+OC_STATIC_ASSERT (
+  OC_BN_WORD_SIZE == 4 || OC_BN_WORD_SIZE == 8,
+  "OC_BN_WORD_SIZE and OC_BN_WORD_NUM_BITS usages must be adapted."
+  );
+
+/**
+  Swaps the byte order of Word.
+
+  @param[in] Word  The Word to swap.
+
+  @returns  The byte-swapped value of Word.
+
+**/
+OC_BN_WORD
+BigNumSwapWord (
+  IN OC_BN_WORD  Word
+  )
+{
+  if (OC_BN_WORD_SIZE == 4) {
```

Comment:
Ugh, better go with sizeof (UINT32) here and everywhere below… 

Also, static analyser like clang-tidy will warn about else, so please write: `if (x) return 1; if (y) return 2;`.

Reply:
Interesting, thanks